lib/host/wasi/inode-win.cpp

```diff
@@ -180,51 +1062,628 @@ WasiExpect<uint64_t> INode::getNativeHandler() const noexcept {
+    }
+
+    wchar_t PathW[MAX_PATH];
+
+    // Convert the path from char_t to wchar_t
+    mbstowcs(PathW, Path.c_str(), MAX_PATH);
+
+    // Append the paths together
+
+    HRESULT CombineResult =
+        PathCchCombine(FullPathW, MAX_PATH, HandleFullPathW, PathW);
+
+    switch (CombineResult) {
+    case S_OK:
+      break;
+    case E_INVALIDARG:
+      return WasiUnexpect(__WASI_ERRNO_INVAL);
+    case E_OUTOFMEMORY:
+      return WasiUnexpect(__WASI_ERRNO_OVERFLOW);
+    default:
+      return WasiUnexpect(__WASI_ERRNO_NAMETOOLONG);
+    }
+
+  } else {
+    mbstowcs(FullPathW, Path.c_str(), MAX_PATH);
+  }
+
+  DWORD AttributeFlags = attributeFlags(OpenFlags, FdFlags);
+  DWORD AccessFlags = accessFlags(FdFlags, VFSFlags);
+  DWORD ShareFlags = shareFlags(OpenFlags);
+  DWORD CreationDisposition = creationDisposition(OpenFlags);
+
+  HANDLE FileHandle = CreateFileW(FullPathW, AccessFlags, ShareFlags, nullptr,
+                                  CreationDisposition, AttributeFlags, nullptr);
+  if (unlikely(FileHandle == INVALID_HANDLE_VALUE)) {
+    return WasiUnexpect(fromWinError(GetLastError()));
+  } else {
+    INode New(FileHandle);
+    return New;
+  }
 }
 
-WasiExpect<void> INode::pathReadlink(std::string, Span<char>,
-                                     __wasi_size_t &) const noexcept {
-  return WasiUnexpect(__WASI_ERRNO_NOSYS);
+WasiExpect<void> INode::pathReadlink(std::string Path, Span<char> Buffer,
+                                     __wasi_size_t &NRead) const noexcept {
+
+  wchar_t FullPathW[MAX_PATH];
+
+  if (PathIsRelativeA(Path.c_str())) {
+    wchar_t HandleFullPathW[MAX_PATH];
+    // First get the paths of the handles
+    if (unlikely(GetFinalPathNameByHandleW(Handle, HandleFullPathW, MAX_PATH,
+                                           FILE_NAME_NORMALIZED)) == FALSE) {
+      return WasiUnexpect(fromWinError(GetLastError()));
+    }
+
+    if (!PathIsDirectoryW(HandleFullPathW)) {
+      return WasiUnexpect(__WASI_ERRNO_NOTDIR);
+    }
+
+    wchar_t PathSuffixW[MAX_PATH];
+
+    // Convert the paths from char_t to wchar_t
+    mbstowcs(PathSuffixW, Path.c_str(), MAX_PATH);
+
+    // Append the paths together
+    HRESULT CombineResult =
+        PathCchCombine(FullPathW, MAX_PATH, HandleFullPathW, PathSuffixW);
+    switch (CombineResult) {
+    case S_OK:
+      break;
+    case E_INVALIDARG:
+      return WasiUnexpect(__WASI_ERRNO_INVAL);
+    case E_OUTOFMEMORY:
+      return WasiUnexpect(__WASI_ERRNO_OVERFLOW);
+    default:
+      return WasiUnexpect(__WASI_ERRNO_NAMETOOLONG);
+    }
+
+  } else {
+    mbstowcs(FullPathW, Path.c_str(), MAX_PATH);
+  }
+
+  // Fill the Buffer with the contents of the link
+  HANDLE LocalFileHandle;
+
+  LocalFileHandle = CreateFileW(FullPathW, GENERIC_READ, FILE_SHARE_READ,
+                                nullptr, OPEN_EXISTING, 0, nullptr);
+
+  if (likely(LocalFileHandle != INVALID_HANDLE_VALUE)) {
+    // MAX_PATH or Buffer.size() in the function call?
```

Comment:
According to the document:
```
If the function succeeds, the return value is the length of the string received by lpszFilePath, in TCHARs. This value does not include the size of the terminating null character.
```

So the size should be the same as the buffer size.